src/lib.rs

```diff
@@ -0,0 +1,389 @@
+//! mdbook-superimport is a pre-processor for [mdbook]'s that helps you avoid link rot
+//! when importing parts of other files into your mdbook.
+//!
+//! [mdbook]: https://github.com/rust-lang-nursery/mdBook
+
+#![deny(missing_docs, warnings)]
+
+use failure::Fail;
+use lazy_static::lazy_static;
+use log::*;
+use mdbook::book::{Book, Chapter};
+use mdbook::preprocess::{Preprocessor, PreprocessorContext};
+use mdbook::BookItem;
+use regex::Regex;
+use std::path::{Path, PathBuf};
+
+/// The pre-processor that powers the mdbook-superimport plugin
+pub struct Superimport;
+
+impl Preprocessor for Superimport {
+    fn name(&self) -> &str {
+        "mdbook-superimport"
+    }
+
+    /// Given a book (usually from stdin) process all of the chapters and replace
+    /// any #superimport's with the content that you're importing.
+    fn run(
+        &self,
+        ctx: &PreprocessorContext,
+        mut book: Book,
+    ) -> Result<Book, mdbook::errors::Error> {
+        debug!("Running `run` method in superimport Preprocessor trait impl");
+
+        let book_src_dir = ctx.root.join(&ctx.config.book.src);
+
+        for section in book.sections.iter_mut() {
+            process_chapter(section, &book_src_dir)?;
+        }
+
+        Ok(book)
```

Comment:
Again for my own learning, is `Ok` like `assert()`? or is it a callback?

Reply:
For sure

It's neither, it's a [Result](https://doc.rust-lang.org/std/result/)

tl;dr 

```
Result {
  Ok(T) <------ this means that everything worked
  Err      <------ there was an error
}
```

Note on line 31 we're returning  a `Result` which is just an enum with two variants `Result::Ok(T)` and `Result::Err`

Reply:
So here we're returning `Ok(book)`